typing/typedecl.ml

```diff
@@ -599,15 +746,187 @@ let check_coherence env loc dpath decl =
+     is given a sort variable.  We may discover this is something more specific
+     while checking the type and other types in the mutually defined group.  If
+     not, we default to value.
+
+     A consequence of this approach is that if you want "any", you have to ask
+     for it.  e.g., in [type 'a foo = Bar], ['a] will get layout [value], but it
+     could be given any.  For that, you need [type ('a : any) foo = Bar].
+
+   - In type kinds, we check that types are representable by unifying them with
+     a sort variable (e.g., arguments to constructors and types used in
+     records). This is enough to ensure we can compile these types.  (Though in
+     the future there will also be the mixed block restriction.)  We default
+     them to value if we don't learn anything by unifying with the type (as may
+     be the case in an existential like [type any = Any : 'a -> any].
+
+   It's important to do this defaulting before things like the separability
+   check or update_decl_layout, because those test whether certain types are
+   void or immediate, and if there were sort variables still around that would
+   have effects!
+*)
+let default_decl_layout decl =
+  (* CR ccasinghino: At the moment, I believe this defaulting is sufficient
+     because of the limited number of places where sort variables are created.
+     But in the future it may be necessary to also do defaulting in the manifest
+     and recursively in the types in the kind, with [iter_type_expr]. *)
+  let default_typ typ =
+    match (Ctype.repr typ).desc with
+    | Tvar { layout } -> Type_layout.default_to_value layout
+    | _ -> ()
+  in
+  let default_ldecl (ldecl : Types.label_declaration) =
+    default_typ ldecl.ld_type
+  in
+  let default_cdecl (cdecl : Types.constructor_declaration) =
+    match cdecl.cd_args with
+    | Cstr_tuple typs -> List.iter default_typ typs
+    | Cstr_record ldecls -> List.iter default_ldecl ldecls
+  in
+  let default_kind = function
+    (* Nothing to do in abstract case because we don't put new sort variables
+       there. *)
+    | Type_abstract _ | Type_open -> ()
+    | Type_record (ldecls, _) -> List.iter default_ldecl ldecls
+    | Type_variant (cdecls, _) -> List.iter default_cdecl cdecls
+  in
+  List.iter default_typ decl.type_params;
+  default_kind decl.type_kind
+
+let default_decls_layout decls =
+  List.iter (fun (_, decl) -> default_decl_layout decl) decls
+
+
+(* Makes sure a type is representable.  Will lower "any" to "value". *)
+(* CR ccasinghino: In the places where this is used, we first call this to
+   ensure a type is representable, and then call [Ctype.type_layout] to get the
+   most precise layout.  These could be combined into some new function
+   [Ctype.type_layout_representable] that avoids duplicated work *)
+let check_representable env loc lloc typ =
+  match Ctype.type_sort env typ with
+  | Ok s -> Type_layout.default_to_value (Sort s)
```

Comment:
I agree!  I'm not sure what I was thinking here.

At some places where this is called, we then perform additional checks (for example: is it void?  is it immediate?) which will have the result of filling in the sort variable incorrectly if we leave it undefaulted.  The right fix is probably to do no defaulting here, and consider what defaulting is needed at each of its call sites.  But in the interest of time, I'm just going to leave a CR here to revisit this - presumably it will need to change in some near future version.